Make examples runnable with `jsaddle-warp`

georgefst opened the pull request:
Allows running examples locally without targeting WASM, e.g. with `cabal run ghc-wasm-miso-examples -- snake`. Note that it doesn't work on Firefox: https://github.com/ghcjs/jsaddle/issues/64.

Code is largely adapted from [Miso's own example](https://github.com/dmjio/miso/blob/master/sample-app-jsaddle/Main.hs). I don't know whether there's any particular reason this hadn't already been done in this repository. Maybe just that GHC's WASM backend is nowhere near as slow as GHCJS, so it's a bit less important? Anyway, I thought I may as well see whether you'd be interested in upstreaming this before going any further.

Note that a major reason for this is HLS support: https://github.com/haskell/haskell-language-server/discussions/4187.

There are still quite a few things that could be improved:

- [ ] The two CPP branches in `Main.hs` are completely separate, so perhaps we're better off using two separate modules and Cabal conditionals instead.
- [ ] The CPP in `App.hs` is a bit awkward, and potentially overly complicated, since `Data.JSString.Internal.Type.JSString` is not the same as `GHC.Internal.Wasm.Prim.Types.JSString`. This is also the reason for the extra annotations in this file and `XHR.hs`: monomorphising the former gives us some compatiblity. It's possible that `jsaddle` could be improved to abstract over this difference.
- [ ] After a non-WASM build, building for WASM with `./build.sh` requires a `cabal clean`. Presumably more could be done to keep build folders separate.
  - I'm not certain yet, but I think this issue has gone away since rebasing on `main` after https://github.com/tweag/ghc-wasm-miso-examples/pull/24.
- [ ] I've ditched the GHCI stuff (i.e. `debug`) from the Miso examples for now. This is simpler and easier to understand, since `jsaddle-warp` has a pretty bad and under-documented API, but it does limit some of the main benefits.
  - It works perfectly well with GHCI (EDIT: except that writing to stdout no longer works...), but when compiled the program immediately just exits.
  - I'm not sure what the purpose of Miso's `debugOr 8080 (f >> syncPoint) jsaddleApp` is. I've found that just using `debug 8000 f` works fine.
- [ ] CSS isn't picked up. We can host the CSS files using `wai-app-static`, replacing `jsaddleApp` with `(jsaddleAppWithJsOr (jsaddleJs False) $ staticApp $ defaultWebAppSettings "frontend")`, but I don't know how we can specify stylesheets like in our static HTML since we can't seem to get at the document's head.
- [ ] For some reason, `fail "unknown example"` isn't called when passing an invalid CLI argument.
- [ ] The XHR example is not yet implemented since it was not trivial to adapt.
- [ ] This needs to be run _outside_ of the `nix develop` shell. I'm not totally sure why, but Cabal otherwise fails in fairly inscrutable ways.

georgefst committed:
Make examples runnable with `jsaddle-warp`

amesgen reviewed (commented):
> Code is largely adapted from Miso's own examples. I don't know whether there's any particular reason this hadn't already been done in this repository. Maybe just that GHC's WASM backend is nowhere near as slow as GHCJS, so it's a bit less important? Anyway, I thought I may as well see whether you'd be interested in upstreaming this before going any further.

Definitely interested, we even had jsaddle-warp support in this repo in the past, but removed it because it caused churn (see https://github.com/tweag/ghc-wasm-miso-examples/pull/6#pullrequestreview-1928769771). Also, the fact that the WASM backend is much faster than GHCJS definitely is a part of the somewhat reduced benefit.

However, recent social media posts like https://redd.it/1g2lanx show that not everyone not aware of this possibility (which in particular gives HLS support), so I think it would be quite nice to demo this in this repo, thanks for (re)starting that effort! Will take a look soon.

georgefst committed:
Use latest Miso

All previous versions are broken, especially without Hackage revisions to add bounds: https://github.com/dmjio/miso/pull/752.

georgefst committed:
WIP - support assets including stylesheets

georgefst committed:
Fix typo

georgefst commented:
This is probably worth rewriting from scratch on top of https://github.com/dmjio/miso/pull/768 and ideally https://github.com/dmjio/miso/pull/770. A lot of the comments in OP are outdated anyway, and the history has become a bit messy as I've experimented with things.

georgefst commented:
Then again, given that Miso's own examples now build with Wasm since https://github.com/dmjio/miso/pull/768, maybe it's time to think about retiring this whole repository...
